### PR TITLE
Sidebar

### PR DESCRIPTION
**Pull request by tristanlatr**
Hello, this PR add a major usability improvement: a left sidebar (#257) that includes : 

- The Table Of Contents if titles are defined in the module/class docstring (only works with reStructuredText, should not be an issue since the plan is to convert epytext element to nodes anyway at some point). It only shows the first 4 level tho: This is arbitrary and I'll probably make it 6. 
- for classes: information about the contents of the current class and parent module/package, sorted by types. 
- for modules/packages: information about the contents of the module and parent package,  sorted by types. 

The sidebar is collapsable in the same fashion docs.python.org/3 does. 
At the same time it is sticky and allows you to scroll it independently of the main page. 

The sidebar can be schematized as follow:
```
+---------------------------
| <Table of contents of the    
| module or class                  
| docstring if titles are 
| defined>
| --------------------------
| <Contents of the current
| module or class >
| --------------------------
| <Contents of the parent 
| of the current
| module or class >
+---------------------------
```

Oh and also it stores user preferences into cookies so we don't have to re-click on collapse sidebar button. 

See example here: https://pydoctor--370.org.readthedocs.build/en/370/api/pydoctor.templatewriter.pages.sidebar.ExpandableItem.html

Details and TODOS:

- New options:
    - --sidebar-expand-depth default is 1
    - --sidebar-toc-depth default is 6
    - --no-sidebar default is False

- TODOs
    -  create an issue one this PR is merged with the following content: “With the side bar generation, especially on project that contain lots of classes per modules and with a value higher than 1 for option --sidebar-expand-depth, files can rapidly take a few megas and be really long to load, even locally. To ensure the best user experience no matter the value of `--sidebar-expand-depth`, we should generate the sidebars for each documentable in a separate file (i.e. `.sidebar.html`) and load it with JS when the rest of the content has been loaded.”

Thanks.

### CODE REVIEW

**Comment by tristanlatr** [comment]
This PR would gain to integrate #360 , The `Package` init module thing is making this more complicated that it has to be. And I believe you fixed it in #360

**Comment by tristanlatr** [comment]
After some tests with really big projects, it seems that we need to have a JS loader that will load the sidebar after the main page content has been loaded. Files can rapidly take a few megas and be really long to load, even locally.

The first step would be to generate the sidebars for each documentable in a separate file and load it with JS when the rest of the content has been loaded.

We could even go further and try to avoid generating the same modules and classes sidebar again and again for each members (with subtle differences in the highlights). Instead we could store the members data as JSON and generate the sibebar from that. But honestly, I don’t think I have the patience and JS knowledges to go that route.

**Comment by tristanlatr** [comment]
>  Files can rapidly take a few megas and be really long to load, even locally.

This is mainly due to the fact that the sidebar contain expandable items, and the default value for `--sidebar-expand-depth` is 3, which is a lot when we're talking about projects with hundreds of classes.

Generating the sidebar HTML with a different file (i.e. `<fullName>.sidebar.html)` and loading it from JS can be done with a little extra effort compared to the second approach. 

So I would be inclined to do that and consider the issue fixed.

**Comment by tristanlatr** [comment]
Do not forget to bump version of the template files when required.

File: .gitignore
Comment (tristanlatr) [inline .gitignore:13]:
```suggestion
```

File: pydoctor/epydoc/markup/plaintext.py
Comment (tristanlatr) [inline pydoctor/epydoc/markup/plaintext.py:52]:
```suggestion
```

File: pydoctor/epydoc/markup/__init__.py
Comment (tristanlatr) [inline pydoctor/epydoc/markup/__init__.py:114]: Maybe it's best to keep it abstract?

File: pydoctor/driver.py
Comment (tristanlatr) [inline pydoctor/driver.py:459]: I should add a test that ensure those warnings are getting trigered.

File: pydoctor/epydoc/markup/epytext.py
Comment (tristanlatr) [inline pydoctor/epydoc/markup/epytext.py:1360]:
```suggestion
```

File: pydoctor/epydoc/markup/__init__.py
Comment (tristanlatr) [inline pydoctor/epydoc/markup/__init__.py:137]: I think it's best to keep it abstract.

File: pydoctor/epydoc2stan.py
Comment (tristanlatr) [inline pydoctor/epydoc2stan.py:850]: This is not tested because the of order of execution of methods (but it's there just in case basically because we never know), utimately we'de like all docstrings to be parsed at the same time, in a specific step of the process.

File: pydoctor/epydoc2stan.py
Comment (tristanlatr) [inline pydoctor/epydoc2stan.py:856]: Why is this untested?

File: pydoctor/templatewriter/pages/__init__.py
Comment (tristanlatr) [inline pydoctor/templatewriter/pages/__init__.py:280]: --nosidebar option should be tested.

File: pydoctor/templatewriter/pages/sidebar.py
Comment (tristanlatr) [inline pydoctor/templatewriter/pages/sidebar.py:208]: Test this.

**Review by tristanlatr — COMMENTED** [review]
A few tests to add...

**Comment by tristanlatr** [comment]
Todo: create an issue one this PR is merged with the following content: “With the side bar generation, especially on project that contain lots of classes per modules and with a value higher than 1 for option `--sidebar-expand-depth` (default is 2), files can rapidly take a few megas and be really long to load, even locally. To ensure the best user experience no matter the value of `--sidebar-expand-depth`, we should generate the sidebars for each documentable in a separate file (i.e. <fullName>.sidebar.html)  and load it with JS when the rest of the content has been loaded.”

**Comment by tristanlatr** [comment]
@adiroiban can you do a review of the design/usability of the table of contents for now ? Thanks

**Comment by adiroiban** [comment]
it looks good. Thanks.

I am not the best person to evaluate API docs tools. I don't use such tools on a daily bases... I just read the code :)

If you are happy, we can have it released and then users can send feedback and improvements and suggestions

**Comment by tristanlatr** [comment]
Maybe it would be good to have a feedback from people that actually use pydoctor interface more often. I don’t know if @glyph or @graingert could give a hand. Thanks.

**Comment by adiroiban** [comment]
Some people might be busy :) 
I guess is best to just release this feature and get more feedback from end users :)

**Comment by tristanlatr** [comment]
@iosonofabio 

While tou are there, can you give a little feedback on the style/usability of the new sidebar ? Would you use it for python-igraph ?

Thanks!

**Comment by tristanlatr** [comment]
@ntamas maybe you have an opinion? Thanks

**Comment by ntamas** [comment]
Seems like a good idea to me! In the mid-term we plan to do away with most of the custom styling we apply on the generated docs and we'll just go for the standard PyDoctor layout (we are not super picky any more about matching the style of the igraph homepage), and in this case this will definitely be useful.

Summon @iosonofabio for a second opinion.

**Comment by iosonofabio** [comment]
I sensed vibrations of my name in the ether and came at once. The example link you posted looks super useful.

I'd wonder about integration with sphinx, readthedocs, and related sidebars though, is that something you want to worry about? As @ntamas says, probably fine for python-igraph as we abandon our branding in favour of doc usability.

**Comment by tristanlatr** [comment]
Thanks for your feedback, to disable the collapsing items, we could set the value of option --sidebar-expand-depth to 1 by default. 

Do you also mean we should not  list the parent members items sidebar ?

**Comment by not-my-profile** [comment]
Ok I am fine with showing the parent members. I don't like the expanding / collapsing at all. I think users would rather just search for something by typing what they're looking for instead of collapsing/expanding stuff in the sidebar. If we can get #381, I don't think we need the collapsing.

Regarding the styling: I think it would look better if the headings in the sidebar were centered. And I'd remove the darker background for the parent section.

I'd also prefer the sidebar to be to the left of where the documentation currently is:

current master:
image

current PR:

image

I'd rather have:

image

**Comment by tristanlatr** [comment]
Collapsing/Expanding was part of the initial motivation of the sidebar, though I get your point. I think we should offer this feature as an option anyway, it can be useful not avid clicking to much around, especially in case where there are inner classes, like in django models.

I’ve tried centered headings and it doesn’t feel right to me.

> And I'd remove the darker background for the parent section.

what would you instead to make it clear it’s not the same section ? 

About the header and the sidebar location, I find it better the way it is now, but I’m not too attached to this, if there are other people that feels the same way as you, I would be inclined to change it. Maybe the readthedocs theme is designed in a better way regarding this specific issue.

**Comment by glyph** [comment]
@tristanlatr OK, looking at this one instead :-) But I see a ton of conflicts here too. Can you resolve them?

File: pydoctor/templatewriter/pages/__init__.py
Comment (tristanlatr) [inline pydoctor/templatewriter/pages/__init__.py:76]: Not sure why passing the template_lookup is required here?

File: pydoctor/themes/base/apidocs.css
Comment (tristanlatr) [inline pydoctor/themes/base/apidocs.css:294]: Is this required ?
```suggestion

```

Th current width is 250px and it's already a bit small...

File: pydoctor/themes/base/apidocs.css
Comment (tristanlatr) [inline pydoctor/themes/base/apidocs.css:305]:
```suggestion
        width: 175px;
```

File: pydoctor/themes/base/common.html
Comment (tristanlatr) [inline pydoctor/themes/base/common.html:8]: I should update pydoctor-template-version

File: pydoctor/themes/base/index.html
Comment (tristanlatr) [inline pydoctor/themes/base/index.html:8]: Update version as well

File: pydoctor/themes/base/nameIndex.html
Comment (tristanlatr) [inline pydoctor/themes/base/nameIndex.html:9]: Update version

File: pydoctor/epydoc/markup/epytext.py
Comment (tristanlatr) [inline pydoctor/epydoc/markup/epytext.py:483]: Special care should be taken to ensure we don't generate twice the same ID for the section header.

File: pydoctor/test/epydoc/test_restructuredtext.py
Comment (tristanlatr) [inline pydoctor/test/epydoc/test_restructuredtext.py:253]: Using the same slugify function as in epytext would produce nicer IDs.

**Comment by tristanlatr** [comment]
I've solve the conflicts and I think we're getting closer. 

@glyph, I you'de like to have a look and leave your review that would be very appreciated. 
Note that it's overhaul missing a few tests, but this part of the code is a bit hard to test :/ I'll see what I ca do. 

Thanks.

File: pydoctor/epydoc/markup/plaintext.py
Comment (tristanlatr) [inline pydoctor/epydoc/markup/plaintext.py:50]: undo this change.

File: pydoctor/themes/base/summary.html
Comment (tristanlatr) [inline pydoctor/themes/base/summary.html:10]: I should update template version here too

File: pydoctor/test/test_node2stan.py
Comment (tristanlatr) [inline pydoctor/test/test_node2stan.py:87]:
```suggestion
    'mailto:postmaster@example.net', 'This is just a note with nested contents']

```

File: pydoctor/epydoc/markup/epytext.py
Comment (tristanlatr) [inline pydoctor/epydoc/markup/epytext.py:481]:
```suggestion
    # Add the section's and heading's DOM elements.
```

**Comment by tristanlatr** [comment]
Ok we have a problem: with the current default configuration,`tox -e cpython-apidocs` took (in the github CI) 4h 2m 37s. So this is a lot of time :/ I think we need to set the default value of --sidebar-expand-depth to 1.

File: pydoctor/options.py
Comment (tristanlatr) [inline pydoctor/options.py:216]:
```suggestion
        '--sidebar-expand-depth', metavar="INT", type=int, default=1, dest='sidebarexpanddepth',
        help=("How many nested modules and classes should be expandable, "
              "first level is always expanded, nested levels can expand/collapse. Value should be 1 or greater. (default: 1)"))
```

**Comment by tristanlatr** [comment]
This PR is good à for a review

**Comment by tristanlatr** [comment]
43m 45s is now needed to generate cpython apidocs with the sidebar (and about 21 mn without). It's the only system test that gain as much of build time proportionnally (about 50% more time). 

For the sake of the speed, I think I'll use option `--no-sidebar` for this system test.

**Comment by glyph** [comment]
> 43m 45s is now needed to generate cpython apidocs with the sidebar (and about 21 mn without). It's the only system test that gain as much of build time proportionnally (about 50% more time).

I apologize for my inability to maintain focus on this fairly complex set of changes, but if it's taking these incredibly huge amounts of time to generate these sidebars, how much *data* are we producing? Loading the pages and clicking around seems good (I am VERY excited to have these available on the public API docs!) but I'm not seeing gigabytes of data getting loaded into my browser, so I'm confused about where these hours and hours of build times are going.

File: pydoctor/epydoc/markup/epytext.py
Comment (tristanlatr) [inline pydoctor/epydoc/markup/epytext.py:1477]: This line might require more check to ensure children is not an empty list?

**Comment by tristanlatr** [comment]
@glyph I think we are good to merge this PR, I'll just release another version of pydoctor before and then let's merge this PR.

**Comment by tristanlatr** [comment]
@glyph, can I get a approval on this PR ?

Thanks

**Comment by tristanlatr** [comment]
@adiroiban maybe ?

**Review by adiroiban — APPROVED** [review]
If you think that this is useful this can be merged.
There is `--no-sidebar` if people don't like it.

And if someone has feedback or suggested changes, it can always be changed in a future PR.

--------

I am not a heave user of API documentation in HTML form.

I am using the Python stdlib docs, but the sidebar over there does something else.

So I can't provide any usability feedback.

Thanks!

File: docs/source/conf.py
Comment (tristanlatr) [inline docs/source/conf.py:110]: `--sidebar-expand-depth=1` is the default anyway now.

```suggestion
```

**Comment by tristanlatr** [comment]
Ok I'm merging this, thanks for reviewing.
Now let's look at #457

**Pull request merged**
